src/protocol/DataFeed.sol

```diff
@@ -13,21 +13,22 @@ contract DataFeed is IDataFeed {
     }
 
     /// @inheritdoc IDataFeed
-    function publish(bytes[] calldata attributes) external override {
+    function publish(bytes[] calldata attributes) external override returns (PublicationHeader memory header) {
         uint256 id = publicationHashes.length;
-        PublicationHeader memory header = PublicationHeader({
+        header = PublicationHeader({
             id: id,
             prevHash: publicationHashes[id - 1],
             publisher: msg.sender,
             timestamp: block.timestamp,
             blockNumber: block.number
         });
 
-        bytes32[] memory attributeHashes = new bytes32[](attributes.length);
-        for (uint256 i; i < attributes.length; ++i) {
+        uint256 nAttributes = attributes.length;
+        bytes32[] memory attributeHashes = new bytes32[](nAttributes);
+        for (uint256 i; i < nAttributes; ++i) {
             attributeHashes[i] = keccak256(attributes[i]);
         }
-        bytes32 pubHash = keccak256(abi.encode(header, attributeHashes));
+        bytes32 pubHash = keccak256(abi.encode(header, keccak256(abi.encode(attributeHashes))));
```

Comment:
hash the attributesHashes first.

Reply:
Just to confirm, is the reason of this change to decouple the attributeHashes to its hash so that someone may decide not to store the list of attributeHashes?